---
title: "07_Barseq_Variance"
author: "Simone Zaghen"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    toc: true
    number_section: true
    toc_float: 
      collapsed: false
      smooth_scroll: true
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "../results/") })

---

Akriti found some interesting genes from the Barseq experiment, and we decided to check the gene variability associated with those genes. She sent me a list of genes with fold change in two timepoints. From that list I will extract the genes with consistently high fold change across timepoints, and I will use those genes to extract information about their variability in the single cell dataset I am working with.

To be noted is that Akriti's data are from fluconazole treatment, while mine are from rapamycin treatment. Also her experimental design and the question she was asking to the dataset were different compared to what i am doing. But we still wanted to see if we had any genes interesting for both that might require further study.

# Load packages

```{r, results='hide', message=FALSE}
library(tidyverse)
library(readxl)
library(kableExtra)
source("functions/LoadWholeWtDataset.R")

options(dplyr.summarise.inform = FALSE)

```

# Load data

Load the full dataset for the wild type strain. The function loadWholeWtDataset takes care of importing data and metadata, normalizing them, and calculating the timebins for each cell in the dataframe (divide cells into 3 minutes timebins). 

```{r, results='hide', message=FALSE}
counts <- loadWholeWtDataset(rawdata_path = "../data/2021_RAPA_TIMECOURSE.tsv.gz",
                             metadata_path = "../data/Cell_Cycle_Metadata_01112024.tsv.gz")
intracluster_output <- readRDS("../results/IntraCluster_AllTimePoints/intracluster_analysis_output.Rds")
barseq <- read_delim("../data/barcode_seq_240223", col_types = cols())

```

Since we don't have many cells per timebin after rapa tratment, we decided to merge sampling timepoints (column Pool) two by two. Then transform df into list (each element is a sampling point) to be able to loop functions through the list. 

```{r}
counts$Pool <- str_replace_all(counts$Pool, c("1" = "2",
                                              "3" = "4", 
                                              "5" = "6", 
                                              "7" = "8"))
```

# Check gene list from barcode sequencing data

First I extract the genes that are interesting from Akriti's gene list. Then I use those genes to filter the dataframes in which I have stored the adjusted rank combined. 

```{r}
#extract interesting genes from barseq data
genes_to_plot <- barseq %>%
  filter(time_48h > 6 &
         time_72h > 6) %>%
  pull(genes)

#extract the stats of scRNAseq with barseq genes
rank_barseq <- intracluster_output %>%
  map(~ pluck(.x, "rank")) %>% #extract nested list
  map2(names(intracluster_output), ~.x %>% #add col with sampling point
         mutate(sampling_point = as.numeric(.y))) %>%
  map(~ filter(.x, genes %in% genes_to_plot))

rm(genes_to_plot)

```

# Plot the rank over sampling points of the barseq genes

For those genes whose ranks extracted, plot how the rank changes over time.

```{r, fig.width=8, fig.height=12}
rank_barseq %>%
    bind_rows() %>%
    mutate(gene = paste(common_name, genes, sep = "|")) %>% #use both genes names
    ungroup() %>%
    select(c("rank_combined","sampling_point", "gene")) %>% 
    mutate(gene = fct_reorder(gene, rank_combined, .desc = T)) %>%
    ggplot(aes(x = sampling_point,
               y = rank_combined,
               color = gene
               )) + 
    geom_line(linewidth = 2) + 
    ylim(0,1) +
    xlab("Sampling Point") +
    theme(legend.position = "none") +
    facet_wrap(~ gene, ncol = 5)

```

# Table with those genes and their infos and stats

For those interesting genes, make a table with all their info (logFC from Akriti's experiment, adjusted rank, and gene function/description).

```{r, results='asis'}
for (i in 1:length(rank_barseq)) {
  print(paste("Sampling point ", names(rank_barseq)[i]))
  cat("\n")
  
  print(rank_barseq[[i]] %>% 
          select(c("genes","common_name","rank_combined", "description")) %>% 
          merge(barseq %>% select(-c("common_name"))) %>%
          relocate(description, .after = time_72h) %>%
          arrange(desc(rank_combined)) %>%
          kable %>%
          kable_styling("striped", full_width = T) %>% 
          scroll_box(height = "500px"))
cat("\n")

}

```

# Plot PDR5, ERG11, FLR1

By looking at the graphs with Chris, we thought that PDR5, ERG11, and FLR1 are interesting, and we want to check their stats more in depth.

First, un-nest the nested list

```{r }
stats_barseq <- intracluster_output %>%
  map(~pluck(.x, "stats"))

for (i in 1:length(stats_barseq)) {
  tmp <- stats_barseq[[i]] %>% 
    map2(names(stats_barseq[[i]]), 
         ~ .x %>% 
           mutate(time = .y)) %>%
    bind_rows()
stats_barseq[[i]] <- tmp
}

stats_barseq <- stats_barseq %>%
  map(~filter(.x,
              common_name %in% c("PDR5",
                                  "ERG11",
                                  "FLR1"))) %>%
  map2(names(stats_barseq),
       ~.x %>%
         mutate(sampling_point = .y)) %>%
  bind_rows() %>%
  mutate(time = as.numeric(time))

rm(tmp)

```

## Plot mean expr and residuals of PDR5, ERG11, FLR1

Some points are missing in the graphs because those timebins didn't have enough cells to compute the statistics.

```{r, fig.height=8, fig.width=8}
genes_to_plot <- c("PDR5", "ERG11", "FLR1")

for (i in 1:length(genes_to_plot)) {
  print(stats_barseq %>%
    filter(common_name %in% genes_to_plot[i]) %>%
    select(c("common_name", "mean", "residuals", "time", "sampling_point")) %>%
    mutate(sampling_point = paste0("Sampling_point_", sampling_point)) %>%
    pivot_longer(cols = c("mean", "residuals"),
                 names_to = "variable",
                 values_to = "value") %>%
    ggplot(aes(x = time, #plot
               y = value)) +
    geom_point() +
    ggtitle(genes_to_plot[i]) +
    expand_limits(x = 0, y = 0) +
    scale_x_continuous(breaks = seq(from = 0,
                                    to = 90,
                                    by = 6)) +
    facet_wrap(sampling_point ~ variable, 
               scales = "free", 
               ncol = 2))
}

```

## Violing plot of PDR5, ERG11, FLR1 counts in each sampling point

```{r, fig.height=8, fig.width=8, fig.show='hold', warning=FALSE}
counts %>%
  select(c("YOR153W", "YHR007C", "YBR008C","Pool")) %>%
  rename(PDR5 = "YOR153W", 
         ERG11 = "YHR007C", 
         FLR1 = "YBR008C") %>%
  pivot_longer(cols = !Pool,
               names_to = "gene",
               values_to = "counts") %>%
  ggplot(aes(x = Pool,
             y = counts)) +
  geom_violin() +
  ggtitle("Linear Scale") +
    facet_wrap(~ gene,
             scales = "free",
             ncol = 1)

counts %>%
  select(c("YOR153W", "YHR007C", "YBR008C","Pool")) %>%
  rename(PDR5 = "YOR153W", 
         ERG11 = "YHR007C", 
         FLR1 = "YBR008C") %>%
  pivot_longer(cols = !Pool,
               names_to = "gene",
               values_to = "counts") %>%
  ggplot(aes(x = Pool,
             y = log10(counts))) +
  geom_violin() +
  ggtitle("Log Scale") +
    facet_wrap(~ gene,
             scales = "free",
             ncol = 1)

```

